src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState_SourceGenerators.cs

```diff
@@ -94,21 +104,26 @@ static SourceGeneratorMap ComputeSourceGenerators(ProjectState projectState)
     public async Task<bool> HasSourceGeneratorsAsync(ProjectId projectId, CancellationToken cancellationToken)
     {
         var projectState = this.SolutionState.GetRequiredProjectState(projectId);
+        if (projectState.AnalyzerReferences.Count == 0)
+            return false;
 
-        if (!s_hasSourceGeneratorsMap.TryGetValue(projectState, out var lazy))
+        if (!RemoteSupportedLanguages.IsSupported(projectState.Language))
+            return false;
+
+        var analyzerReferenceMap = s_languageToAnalyzerReferenceMap[projectState.Language];
+        if (!analyzerReferenceMap.TryGetValue(projectState.AnalyzerReferences, out var lazy))
         {
             // Extracted into local function to prevent allocations in the case where we find a value in the cache.
```

Comment:
>// Extracted into local function to prevent allocations in the case where we find a value in the cache. [](http://example.com/codeflow?start=12&length=103)

Could we just make static and get rid of this comment?

Reply:
my guess is no.  there are needed captures in these paths.  but i will try.

Reply:
yeah, it's all sorts of captury.  :)  No go on static here.  We need to capture enough data to be able to call to oop effectively to answer the question. note that once this happens once, and is cached, that captured data will be dropped.  So we only need this alloc when looking at a fresh (non-identity) list of analyzer-references.  Which should be very rare as the solution forks forwards.